Python adapter crashes Neovim

Hello,

I have tried to follow the examples given in the manual and I get very weird behaviour with the `debugpy` adapter.

## My configuration

```lua
dap.adapters.python = {
	type = 'executable',
	command = '/usr/bin/python3',
	args = {'-m', 'debugpy.adapter'}
}

dap.configurations.python = {
	{
		type = 'python',
		request = 'launch',
		name = 'Launch file',
		program = "${file}",
		pythonPath = function(adapter)
			return '/usr/bin/python3'
		end,
	}
}
```

## The code

Here is the Python script I want to debug

```python
#!/bin/env python3

def test_fun(x, y):
    print('This is a test function')
    print('The value of x is ' + str(x))
    print('The value of y is ' + str(y))
    sum = x + y
    print('The sum is ' + str(sum))
    print('bye')

test_fun(2, 3)
```

Whether the file is executable or not does not matter.


## What goes wrong

I set a break point on line 6, open a debug REPL, then start the debugger with `lua require"dap".continue()`. The debugger stops on the break point and the REPL shows the printed lines. Everything good so far.

However, once I press any key, even if it is just to move the cursor, Neovim gets suspended as if I had pressed `CTRL-Z`. If I try resuming it with `fg` everything is totally borked for lack of a better word, and I can only `CTRL-C` to quit.

I have not been able to test other adapters, I wanted to get this one working first because it's the easiest one to set up. While I'm at it, what are the keys to use in the `dap.configurations.<whatever>` table? Do they depend on the particular adapter?

I guess a adapter definition like the following should work:

```lua
dap.adapters.kotlin = {
  type = 'executable';
  command = os.getenv('HOME') .. '/dev/fwcd/kotlin-debug-adapter/adapter/build/install/adapter/bin/kotlin-debug-adapter';
}
```

Where the `command` path is adapted.

For the configuration you could look at the `launch.json` files, like this one https://github.com/fwcd/kotlin-debug-adapter/blob/c01e2f950f77e6e639063099ae1754e109bc1875/.vscode/launch.json#L11-L17 


I tried to use the adapter, but when I start debugging I get some errors:

```
[ERROR] server    Internal error: java.lang.reflect.InvocationTargetException
[ERROR] java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
[ERROR] 	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:67)
[ERROR] 	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.request(GenericEndpoint.java:120)
[ERROR] 	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleRequest(RemoteEndpoint.java:261)
```

Looks like a kotlin-debug-adapter issue to me.


Regarding the original python adapter crash issue:

Which terminal are you using?
Your example script isn't producing any weird escape sequences, so I can't image why it locks up.

Could you maybe share a gif or so that shows how it acts up?
Could any other plugins be interfering?

--- a/lua/dap.lua
+++ b/lua/dap.lua
@@ -859,6 +859,7 @@ function Session:spawn(adapter, opts)
     stdio = {stdin, stdout, stderr};
     cwd = options.cwd;
     env = options.env;
+    detached = true;
   }, onexit)
   assert(handle, 'Error running ' .. adapter.command .. ': ' .. pid_or_err)
   o.client = {
@@ -1348,4 +1349,13 @@ function M.set_log_level(level)
 end
 
 
+function M._vim_exit_handler()
+  if session then
+    session:close()
+  end
+end
+dap._vim_exit_handler = M._vim_exit_handler  -- luacheck: ignore 112
+
+
+api.nvim_command("autocmd VimLeavePre * lua dap._vim_exit_handler()")
 return M

